File: ios/MedicationApp/Features/Today/PatientTodayNextSlotSelector.swift
import Foundation

struct PatientTodayNextSlotSelector {
    struct Candidate {
        let slot: NotificationSlot
        let scheduledAt: Date
        let remainingCount: Int
        let isWithinRecordingWindow: Bool
        let isLate: Bool
        let hasRecordableInventory: Bool
    }

    static func selectSlot(from candidates: [Candidate], now: Date = Date()) -> NotificationSlot? {
        candidates
            .filter { candidate in
                candidate.remainingCount > 0
                    && candidate.hasRecordableInventory
                    && (
                        candidate.scheduledAt >= now
                            || (candidate.isWithinRecordingWindow && !candidate.isLate)
                    )
            }
            .sorted { lhs, rhs in
                lhs.scheduledAt < rhs.scheduledAt
            }
            .first?
            .slot
    }
}

File: android/app/src/main/java/com/afterlifearchive/medmanager/data/patient/PatientTodayNextSlotSelector.kt
package com.afterlifearchive.medmanager.data.patient

import java.time.Instant

data class NextSlotCandidate(
    val slot: MedicationSlot,
    val scheduledAt: Instant,
    val remainingCount: Int,
    val isWithinRecordingWindow: Boolean,
    val hasRecordableInventory: Boolean,
    val isLate: Boolean = false,
)

object PatientTodayNextSlotSelector {
    fun select(candidates: List<NextSlotCandidate>, now: Instant = Instant.now()): MedicationSlot? =
        candidates.asSequence()
            .filter { candidate ->
                candidate.remainingCount > 0 &&
                    candidate.hasRecordableInventory &&
                    (!candidate.scheduledAt.isBefore(now) ||
                        (candidate.isWithinRecordingWindow && !candidate.isLate))
            }
            .minByOrNull(NextSlotCandidate::scheduledAt)
            ?.slot
}
